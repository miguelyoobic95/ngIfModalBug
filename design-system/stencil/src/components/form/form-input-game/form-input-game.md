---
name: Game
category: Form Components
---

```yoo-form-input-game.html
    <yoo-form-input-game></yoo-form-input-game>
```

```yoo-form-input-game.js
    var comp = document.querySelector('yoo-form-input-game');
    comp.name = 'Runner';
    comp.phaser = Phaser;
```

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`name`|`string`|   |Name of the game you want to play (See the section below for a list of all available games)|
|`phaser`|`any`|   |Phaser instance imported by the framework|
|`fieldId`|`string`|`game-div`|Id of the div where the phaser game will be set up. Leave default value if you don't need a custom value|



## Events
|Attr|Description|
|---|---|
|`gameOver`|Triggered when the game is over|

## Available Game

|Game name| Description|
|---------|--------|
|`Runner`|Game example of Mario|
|`Card`|Card Game|
|`Example`|Example game description|