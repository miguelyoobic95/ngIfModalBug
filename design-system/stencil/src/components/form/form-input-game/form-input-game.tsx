import { Component, Prop, Event, EventEmitter, State, Element } from '@stencil/core';
import { IAbstractGameEntry } from '@shared/interfaces';

import * as fromGame from '../../../games';

@Component({
    tag: 'yoo-form-input-game',
    styleUrl: 'form-input-game.scss',
    scoped: true
})
export class YooFormInputGameComponent {

    @Prop() name: string = fromGame.games.runner;
    @Prop() phaser: any; // Phaser instance imported by the framework
    @Prop() fieldId: string = 'game-div';

    @Event() gameOver: EventEmitter<boolean>;

    @State() isGameOver: boolean;

    @Element() host: HTMLStencilElement;

    private gameProps: IAbstractGameEntry = null; // test purpose
    private game: fromGame.AbstractGame;

    setGameProps(): void {
        this.gameProps = {fieldId: this.fieldId};
        // console.log('gameProps', this.gameProps);
    }

    init(): void {
        this.game = fromGame.gameFactory(this.name);
        this.game.gameover.subscribe(val => {
            this.isGameOver = val;
          });
        // console.log('game', this.game);
        this.setGameProps();
        this.game.initWithPhaser(this.gameProps, this.phaser);
    }

    componentDidLoad() {
        setTimeout(() => this.init(), 300);
    }

    componentWillUpdate() {
        // console.log('Comp Phaser will Upload, isGameOver?, event emitted by Stencil !', this.isGameOver);
        if (this.isGameOver) {
          this.gameOver.emit(true);
        }
      }

    render(): JSX.Element {
        return (
            <div>
                <div class="container" id={this.fieldId}></div>
            </div>
        );
    }
}
