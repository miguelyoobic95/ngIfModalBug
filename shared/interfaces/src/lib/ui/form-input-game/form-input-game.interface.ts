export interface IAbstractGameEntry {
    // common props of phaser games needed in constructor
    fieldId: string;
    assetDir?: string;
}

export interface IInputGameEntry {
    name: string;
    props: IAbstractGameEntry;
}