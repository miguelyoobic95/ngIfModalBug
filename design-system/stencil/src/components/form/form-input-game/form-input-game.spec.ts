import { TestWindow } from '@stencil/core/testing';
import { YooFormInputGameComponent } from './form-input-game';

describe('YooFormInputGameComponent', () => {
  let window;
  beforeEach(() => {
    window = new TestWindow();
  });
  it('Should build', () => {
      expect(new YooFormInputGameComponent()).toBeTruthy();
  });

  describe('Rendering', () => {
      it ('Should render', async () => {
          let element = await window.load({
              components: [YooFormInputGameComponent],
              html: '<yoo-form-input-game></yoo-form-input-game>'
            });
          expect(element).toMatchSnapshot();
      });

      it('Should render with a name prop', async () => {
          let element = await window.load({
              components: [YooFormInputGameComponent],
              html: '<yoo-form-input-game name="Runner" ></yoo-form-input-game>'
            });
          expect(element).toMatchSnapshot();
      });

      it('Should render with a name prop', async () => {
          let element = await window.load({
              components: [YooFormInputGameComponent],
              html: '<yoo-form-input-game field-id="game-field-id" name="Runner" ></yoo-form-input-game>'
            });
          expect(element.fieldId).toEqual('game-field-id');
          expect(element).toMatchSnapshot();
      });

      it('Should receive phaser as a prop', async () => {
          let element = await window.load({
              components: [YooFormInputGameComponent],
              html: '<yoo-form-input-game phaser="phaser"></yoo-form-input-game>'
            });
          expect(element.phaser).toEqual('phaser');
          expect(element).toMatchSnapshot();
      });
  });
  describe('events', () => {
    it ('should emit gameOver', async () => {
      let element = new YooFormInputGameComponent();
      element.gameOver = {
        emit: () => {}
      };
      const changedSpy = jest.spyOn(element.gameOver, 'emit');
      element.isGameOver = true;
      element.componentWillUpdate();
      expect(changedSpy).toHaveBeenCalledWith(true);
    });
  });
});
