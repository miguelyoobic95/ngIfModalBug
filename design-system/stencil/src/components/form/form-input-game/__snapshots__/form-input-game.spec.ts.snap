// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooFormInputGameComponent Rendering Should receive phaser as a prop 1`] = `
<yoo-form-input-game
  class="hydrated"
  data-yoo-form-input-game-host=""
  phaser="phaser"
>
  
  <div
    data-yoo-form-input-game=""
  >
    <div
      class="container"
      data-yoo-form-input-game=""
      id="game-div"
    />
  </div>
</yoo-form-input-game>
`;

exports[`YooFormInputGameComponent Rendering Should render 1`] = `
<yoo-form-input-game
  class="hydrated"
  data-yoo-form-input-game-host=""
>
  
  <div
    data-yoo-form-input-game=""
  >
    <div
      class="container"
      data-yoo-form-input-game=""
      id="game-div"
    />
  </div>
</yoo-form-input-game>
`;

exports[`YooFormInputGameComponent Rendering Should render with a name prop 1`] = `
<yoo-form-input-game
  class="hydrated"
  data-yoo-form-input-game-host=""
  name="Runner"
>
  
  <div
    data-yoo-form-input-game=""
  >
    <div
      class="container"
      data-yoo-form-input-game=""
      id="game-div"
    />
  </div>
</yoo-form-input-game>
`;

exports[`YooFormInputGameComponent Rendering Should render with a name prop 2`] = `
<yoo-form-input-game
  class="hydrated"
  data-yoo-form-input-game-host=""
  field-id="game-field-id"
  name="Runner"
>
  
  <div
    data-yoo-form-input-game=""
  >
    <div
      class="container"
      data-yoo-form-input-game=""
      id="game-field-id"
    />
  </div>
</yoo-form-input-game>
`;
